hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/TypeUtils.java

```diff
@@ -0,0 +1,328 @@
+            return fieldTypeName + "CompactType";
+        }
+
+        @Override
+        protected String getSchemaId(Map<String, String> options, Boolean isKey) {
+            return options.get(isKey == null ? SqlConnector.OPTION_TYPE_COMPACT_TYPE_NAME : isKey
+                    ? SqlConnector.OPTION_KEY_COMPACT_TYPE_NAME : SqlConnector.OPTION_VALUE_COMPACT_TYPE_NAME);
+        }
+    }
+
+    private static class JavaEnricher extends FieldEnricher<Class<?>, SortedMap<String, Class<?>>> {
+        JavaEnricher(RelationsStorage relationsStorage) {
+            super(QueryDataType.OBJECT_TYPE_KIND_JAVA, relationsStorage);
+        }
+
+        @Override
+        protected String getTypeMetadata(Class<?> typeClass) {
+            return typeClass.getName();
+        }
+
+        @Override
+        protected SortedMap<String, Class<?>> getSchema(Class<?> typeClass) {
+            return FieldsUtil.resolveClass(typeClass);
+        }
+
+        @Override
+        protected List<TypeField> resolveFields(SortedMap<String, Class<?>> classFields) {
+            return classFields.entrySet().stream().map(e -> {
+                if (isUserClass(e.getValue())) {
+                    throw QueryException.error("Column list is required to create nested fields");
+                }
+                return new TypeField(e.getKey(), QueryDataTypeUtils.resolveTypeForClass(e.getValue()));
+            }).collect(toList());
+        }
+
+        @Override
+        protected Class<?> getFieldSchemaId(SortedMap<String, Class<?>> classFields,
+                                            String fieldName, String fieldTypeName) {
+            return classFields.get(fieldName);
+        }
+
+        @Override
+        protected Class<?> getSchemaId(Map<String, String> options, Boolean isKey) {
+            String className = options.get(isKey == null ? SqlConnector.OPTION_TYPE_JAVA_CLASS :
+                    isKey ? SqlConnector.OPTION_KEY_CLASS : SqlConnector.OPTION_VALUE_CLASS);
+            return className != null ? loadClass(className) : null;
+        }
+
+        private static boolean isUserClass(Class<?> clazz) {
+            return !clazz.isPrimitive() && !clazz.getPackage().getName().startsWith("java.");
+        }
+
+        private static Class<?> loadClass(String className) {
+            try {
+                return ReflectionUtils.loadClass(className);
+            } catch (Exception e) {
+                throw QueryException.error("Unable to load class '" + className + "'", e);
+            }
+        }
+    }
+
+    private static class AvroEnricher extends FieldEnricher<Schema, Schema> {
+        AvroEnricher(RelationsStorage relationsStorage) {
+            super(QueryDataType.OBJECT_TYPE_KIND_AVRO, relationsStorage);
+        }
+
+        @Override
+        protected String getTypeMetadata(Schema schema) {
+            // Used only for debugging purposes since AvroUpsertTarget has already a reference to the schema.
```

Comment:
this will consume additional memory and be executed each time a query is executed which might matter for simple queries. BTW, stringified schema is available in mapping options.

Reply:
Only the stringified version of the topmost schema is in mapping options.

Reply:
is it really that beneficial? it has some cost and we might accidentally start using it. It will be stored in `__sql.catalog`.
If this is sometimes useful, add a property to enable this.

Reply:
It is only computed during `CREATE MAPPING`. We use type metadata in `KvMetadataResolver`s to resolve missing schemas. (Expect for `KvMetadataAvroResolver`, in which, type name is assumed to be record name. We might allow overriding record name via new option: `avroRecordName`.) We also use type metadata in `UpsertTarget`s to obtain the actual schema in formats that support polymorphism such as Java and Compact. In Portable and Avro, the schemas pointed by type metadata and parent schema are expected to be same.

Reply:
> It is only computed during CREATE MAPPING

but deserialized each time `Mapping` needs to be deserialized and it might be quite big blob of json

Reply:
It is deserialized to `String`, not `org.apache.avro.Schema`.

Reply:
Since it doesn't have any use and has potentially high networking and serialization costs, I am removing type metadata for Avro and also leaving an explanation about the self-containment of Avro schemas.